cardano-node-emulator/src/Cardano/Node/Emulator/MTL.hs

```diff
@@ -0,0 +1,221 @@
+  , handleChain
+) where
+
+import Control.Lens (alaf, makeLenses, view, (%~), (&), (^.))
+import Control.Monad (void)
+import Control.Monad.Error.Class (MonadError, throwError)
+import Control.Monad.Except (ExceptT)
+import Control.Monad.Freer (Eff, Member, interpret, run, type (~>))
+import Control.Monad.Freer.Extras (raiseEnd)
+import Control.Monad.Freer.Extras.Log (LogMessage (..), LogMsg (..))
+import Control.Monad.Freer.State (State, modify, runState)
+import Control.Monad.Freer.Writer qualified as F (Writer, runWriter, tell)
+import Control.Monad.Identity (Identity)
+import Control.Monad.RWS.Class (MonadRWS, ask, get, put, tell)
+import Control.Monad.RWS.Strict (RWST)
+import Data.Foldable (toList)
+import Data.Map (Map)
+import Data.Map qualified as Map
+import Data.Monoid (Endo (..), Sum (..))
+import Data.Sequence (Seq)
+import Data.Text qualified as Text
+import Prettyprinter qualified as Pretty
+import Prettyprinter.Render.Text qualified as Pretty
+
+import Cardano.Api qualified as C
+import Cardano.Api.Shelley qualified as C
+import Cardano.Node.Emulator qualified as E
+import Ledger (CardanoAddress, CardanoTx, DatumFromQuery, DatumHash, DecoratedTxOut, OnChainTx (..),
+               PaymentPrivateKey (..), ToCardanoError, TxOut (..), TxOutRef, UtxoIndex, ValidationErrorInPhase,
+               mkDecoratedTxOut)
+import Ledger.AddressMap qualified as AM
+import Ledger.Index (UtxoIndex (..), createGenesisTransaction, insertBlock)
+import Ledger.Tx (CardanoTx (..), DatumFromQuery (..), addCardanoTxSignature, decoratedTxOutValue, getCardanoTxId)
+import Ledger.Tx.CardanoAPI (CardanoBuildTx (..), fromCardanoReferenceScript, fromCardanoScriptData,
+                             toCardanoTxOutValue)
+import Plutus.V2.Ledger.Api qualified as PV2
+import PlutusTx.Builtins qualified as PlutusTx
+
+
+data EmulatorState = EmulatorState
+  { _esChainState :: E.ChainState
+  , _esAddressMap :: AM.AddressMap
+  }
+  deriving (Show)
+
+makeLenses 'EmulatorState
+
+data EmulatorError
+  = BalancingError E.BalancingError
+  | ValidationError ValidationErrorInPhase
+  | ToCardanoError ToCardanoError
+  deriving (Show)
+
+type MonadEmulator m = (MonadRWS E.Params (Seq E.ChainEvent) EmulatorState m, MonadError EmulatorError m)
+type EmulatorT m = ExceptT EmulatorError (RWST E.Params (Seq E.ChainEvent) EmulatorState m)
+type EmulatorM = EmulatorT Identity
+
+emptyEmulatorState :: EmulatorState
+emptyEmulatorState = EmulatorState E.emptyChainState mempty
+
+emptyEmulatorStateWithInitialDist :: Map CardanoAddress C.Value -> EmulatorState
+emptyEmulatorStateWithInitialDist initialDist =
+  let tx = Valid $ createGenesisTransaction initialDist
+  in emptyEmulatorState
+    & esChainState . E.chainNewestFirst %~ ([tx] :)
+    & esChainState . E.index %~ insertBlock [tx]
+    & esAddressMap %~ AM.updateAllAddresses tx
+
+handleChain :: MonadEmulator m => Eff [E.ChainControlEffect, E.ChainEffect] a -> m a
+handleChain eff = do
```

Comment:
It could be useful to test these handlers. Maybe in another story?

Reply:
It is used by `queueTx` and `nextSlot`, and they are tested.